### PR TITLE
Switch the virtual-char system to use segmented lists.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**

### CODE REVIEW

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/EmbeddedLanguages/VirtualChars/AbstractVirtualCharService.cs [hunk 1]
```diff
@@ -221,7 +223,7 @@ protected static bool IsOpenOrCloseBrace(char ch)
                 return sequence.GetSubSequence(TextSpan.FromBounds(startIndexInclusive, endIndexExclusive));
             }
 
-            return VirtualCharSequence.Create(result.ToImmutable());
+            return VirtualCharSequence.Create(result.ToImmutableSegmentedList());
```
Comment (sharwell) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/EmbeddedLanguages/VirtualChars/AbstractVirtualCharService.cs:226]:
```suggestion
            return VirtualCharSequence.Create(result.ToImmutable());
```

File: src/Dependencies/Collections/SegmentedList`1.cs [hunk 2]
```diff
@@ -308,6 +308,9 @@ public int BinarySearch(T item)
         public int BinarySearch(T item, IComparer<T>? comparer)
             => BinarySearch(0, Count, item, comparer);
 
+        internal int BinarySearch<TValue>(TValue value, Func<T, TValue, int> comparer)
+            => SegmentedArray.BinarySearch(_items, value, comparer);
```
Comment (sharwell) [inline src/Dependencies/Collections/SegmentedList`1.cs:312]: ❗ This failed to account for cases where `Capacity < Count`

Comment (CyrusNajmabadi) [reply src/Dependencies/Collections/SegmentedList`1.cs:312]: oooh..

**Comment by CyrusNajmabadi** [comment]
Closing out in favor of https://github.com/dotnet/roslyn/pull/59503/files

**Pull request closed without merging**